public/app/features/explore/utils/supplementaryQueries.ts

```diff
@@ -129,98 +127,66 @@ const getSupplementaryQueryFallback = (
   explorePanelData: Observable<ExplorePanelData>
 ): Observable<DataQueryResponse> | undefined => {
-  if (hasSupplementaryQuerySupport(datasourceInstance, type)) {
-    return datasourceInstance.getDataProvider(type, request);
-  } else if (datasourceInstance.meta?.mixed === true) {
-    const queries = request.targets.filter((t) => {
-      return t.datasource?.uid !== MIXED_DATASOURCE_NAME;
-    });
-    // Build groups of queries to run in parallel
-    const sets: { [key: string]: DataQuery[] } = groupBy(queries, 'datasource.uid');
-    const mixed: Array<{ datasource: Promise<DataSourceApi>; targets: DataQuery[] }> = [];
-
-    for (const key in sets) {
-      const targets = sets[key];
-      mixed.push({
-        datasource: getDataSourceSrv().get(targets[0].datasource, request.scopedVars),
-        targets,
-      });
+  const providers = datasources.map(({ datasource, targets }, i) => {
+    const dsRequest = cloneDeep(request);
+    dsRequest.requestId = `${dsRequest.requestId || ''}_${i}`;
+    dsRequest.targets = targets;
+
+    if (hasSupplementaryQuerySupport(datasource, type)) {
+      return datasource.getDataProvider(type, dsRequest);
+    } else {
+      return getSupplementaryQueryFallback(type, explorePanelData, targets, datasource.name);
     }
+  });
 
-    return from(mixed).pipe(
-      mergeMap((query, i) => {
-        return from(query.datasource).pipe(
-          mergeMap((ds) => {
-            const dsRequest = cloneDeep(request);
-            dsRequest.requestId = `mixed-${type}-${i}-${dsRequest.requestId || ''}`;
-            dsRequest.targets = query.targets;
-
-            if (hasSupplementaryQuerySupport(ds, type)) {
-              const dsProvider = ds.getDataProvider(type, dsRequest);
-              if (dsProvider) {
-                // 1) It provides data for current request - use the provider
-                return dsProvider;
-              } else {
-                // 2) It doesn't provide data for current request -> return nothing
-                return of({
-                  data: [],
-                  state: LoadingState.NotStarted,
-                });
-              }
-            } else {
-              // 3) Data source doesn't support the supplementary query -> use fallback
-              // the fallback cannot determine data availability based on request, it
-              // works on the results once they are available so it never uses the cache
-              return getSupplementaryQueryFallback(type, explorePanelData, query.targets, ds.name);
-            }
-          })
-        );
-      }),
-      scan<DataQueryResponse, DataQueryResponse>(
-        (acc, next) => {
-          if (acc.error || next.state === LoadingState.NotStarted) {
-            return acc;
-          }
+  const definedProviders: Array<Observable<DataQueryResponse>> = providers.filter(
+    (item: Observable<DataQueryResponse> | undefined): item is Observable<DataQueryResponse> => {
+      return !!item;
+    }
+  );
```

Comment:
you can simply use `isTruthy` from `@grafana/data`, it should yield the same result/types:

```suggestion
  const definedProviders = providers.filter(isTruthy);
```

Reply:
Sweet! Exactly what I was looking for. Changed in c21cee1cb0add4cb177c56f4c41dd2d7551d3203.